File: get_jwt.py
# get_jwt.py
#
# Usage: python get_jwt.py <KEYPAIR_JSON_PATH>
# Example: python get_jwt.py rugcheck-keypair.json

import sys, json, time
import base58
import nacl.signing
import requests

def main():
    if len(sys.argv) != 2:
        print("Usage: python get_jwt.py <KEYPAIR_JSON_PATH>")
        sys.exit(1)

    keypath = sys.argv[1]
    # 1) Load the Solana keypair JSON (64‐byte array of ints)
    with open(keypath, "r") as f:
        raw = json.load(f)
    full_secret = bytes(raw)  # 64 bytes total

    # 2) Extract only the first 32 bytes as the Ed25519 seed
    seed = full_secret[:32]
    signing_key = nacl.signing.SigningKey(seed)

    # 3) Derive the public key (last 32 bytes from Solana keypair should match this)
    verify_key = signing_key.verify_key
    pubkey_bytes = verify_key.encode()
    pubkey_b58 = base58.b58encode(pubkey_bytes).decode()

    # 4) Sign exactly the literal bytes of "Sign-in to Rugcheck.xyz"
    static_msg = b"Sign-in to Rugcheck.xyz"
    signature_bytes = signing_key.sign(static_msg).signature  # 64 raw bytes

    # 5) Build the 'message' object (with timestamp) – this is NOT what we sign, it's just for the payload
    timestamp = int(time.time())
    msg_obj = {
        "message": "Sign-in to Rugcheck.xyz",
        "publicKey": pubkey_b58,
        "timestamp": timestamp
    }

    # 6) Construct the full authentication payload that RugCheck expects
    payload = {
        "message": msg_obj,
        "signature": {
            "data": list(signature_bytes),  # list of 64 ints 0–255
            "type": "ed25519"
        },
        "wallet": pubkey_b58
    }

    # 7) POST to the RugCheck auth endpoint
    url = "https://api.rugcheck.xyz/auth/login/solana"
    resp = requests.post(url, json=payload, timeout=10)
    if not resp.ok:
        print("Failed to get JWT:", resp.status_code, resp.text)
        sys.exit(1)

    token = resp.json().get("token")
    print("Your RugCheck JWT:", token)

if __name__ == "__main__":
    main()

File: get_jwt.js
// get_jwt.js
//
// Usage: node get_jwt.js <KEYPAIR_PATH>
// Example: node get_jwt.js ./rugcheck-keypair.json

import fs from "fs";
import fetch from "node-fetch";
import nacl from "tweetnacl";
import bs58 from "bs58";

async function main() {
  const keypairPath = process.argv[2];
  if (!keypairPath) {
    console.error("Usage: node get_jwt.js <KEYPAIR_PATH>");
    process.exit(1);
  }

  // 1) Load your Solana keypair JSON (array of 64 numbers)
  const raw = JSON.parse(fs.readFileSync(keypairPath, "utf8"));
  const secretKey = Uint8Array.from(raw);

  // 2) Extract the public key bytes (last 32 bytes of secretKey)
  const publicKeyBytes = secretKey.slice(32);
  const publicKeyBase58 = bs58.encode(publicKeyBytes);

  // 3) Build and sign the fixed message
  const messageToSign = "Sign-in to Rugcheck.xyz";
  const encodedMessage = new TextEncoder().encode(messageToSign);
  const signatureBytes = nacl.sign.detached(encodedMessage, secretKey);

  // 4) Construct the payload
  const timestamp = Math.floor(Date.now() / 1000);
  const body = {
    message: {
      message: messageToSign,
      publicKey: publicKeyBase58,
      timestamp: timestamp
    },
    signature: {
      data: Array.from(signatureBytes),
      type: "ed25519"
    },
    wallet: publicKeyBase58
  };

  // 5) POST to RugCheck’s auth endpoint
  const resp = await fetch("https://api.rugcheck.xyz/auth/login/solana", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body)
  });

  if (!resp.ok) {
    console.error("Failed to get JWT:", resp.status, await resp.text());
    process.exit(1);
  }

  const { token } = await resp.json();
  console.log("Your RugCheck JWT:", token);
}

main().catch((err) => {
  console.error("Error:", err);
  process.exit(1);
});